### PR TITLE
Fix enable/disable toggle button action

### LINKED ISSUE #14: Enable/disable toggle doesn't work as it should.
Howdy!

The enable/disable toggle here doesn't work: 

Screen Shot 2020-02-12 at 10 12 29

If disabled here, it would still show up as a payment method in the checkout as it has no affect with the setting here: 

Screen Shot 2020-02-12 at 10 12 18

The user will need to go into the Payfast settings to disable it there. 

**To Reproduce**

1. Install, activate and configure Payfast
2. Go to wp-admin/admin.php?page=wc-settings&tab=checkout page and toggle Payfast to "disable"
3. Refresh the page. 
4. It will show that Payfast is disabled, but it will still be enabled under wp-admin/admin.php?page=wc-settings&tab=checkout&section=payfast page

**Expected behavior**

I would expect the same behavior to apply to Payfast as it does with the other gateway methods on that page and the toggle to work. 

**Environment (please complete the following information):**

 - WordPress Version 5.3.2
 - WooCommerce Version 3.9.1
 - PayFast Plugin Version 1.4.14
 - Storefront theme Version 2.5.3

### PR DESCRIPTION
**Pull request by tpaksu**
Fixes #14

#### Changes proposed in this Pull Request

Fix enable/disable toggle button action

The status toggle of this plugin was also checking if the plugin is valid for use ( if the merchant ID and key is set, and if the store currency is among the available currencies which can be used with PayFast), but the inner checkbox was only checking if the plugin is enabled or disabled. So while the toggle on the payment providers list was working correct on the backend, it wasn't displaying the correct status of the plugin. 

Also the plugin has 3 indicators that show if the plugin is active or not, and needs to be sync'ed all the time, but wasn't since the first one was also dependent of the settings validity:

```php
$gateway->enabled
$gateway->settings[ 'enabled' ]
$gateway->get_option( 'enabled ')
```

Instead of binding the gateway settings validation to `$gateway->enabled` flag, I'll move the check and the error to a more sensible place, like setting all the "enabled" flags to false if any settings are not valid for usage.

This PR will provide that.

#### Testing instructions

**With valid settings**

- Open up two browser tabs
- From tab #1 navigate to http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout
- From tab #2 navigate to http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=wc_gateway_payfast
- Change the setting from the first tab and click "Save changes" button
- Refresh the second tab and check if it is updated to the setting in the first tab
- Do the reverse, change the setting in the second tab and save, and check if it updates the first one when refreshed.

**With invalid settings**

- Check that you can't enable the payment gateway and see an error related to the missing/invalid setting.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

### CODE REVIEW

**Comment by tpaksu** [comment]
@dwainm I can't set this PR to `needs review` state, because most of the labels are missing! 😁 The "Sync Woo Labels to Repository" workflow source repository needs an update I suppose. I'll move it in ZH without the labels for now.

**Review by ismaeldcom — DISMISSED** [review]
Thanks for working on this issue!

When I try to access [PayFast Setting](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=wc_gateway_payfast) receive the following error:
```
Fatal error: Uncaught Error: Call to undefined method WC_Gateway_PayFast::is_valid_for_use() in /var/repositories/woocommerce-gateway-payfast/includes/class-wc-gateway-payfast-blocks-support.php on line 33
```
This is cause `class-wc-gateway-payfast-blocks-support.php` is calling `is_valid_for_use` that is removed now.

I see some improvements about better admin notices in your solution, but maybe they are a bit out of the scope here. Might be better to fix the issue and then propose those improvements in another one. What do you think?

On the other hand, something like the following change in the [line 73](https://github.com/woocommerce/woocommerce-gateway-payfast/blob/8d4b0fbf51e4c6203eed178a72f33844bfc7558d/includes/class-wc-gateway-payfast.php#L73) could solve the issue.
```php
$this->enabled = $this->is_valid_for_use() ? $this->get_option( 'enabled' ): 'no';
```

**Comment by tpaksu** [comment]
Thank you @ismaeldcom , I've missed that reference. I added a quick fix for it and check for other stuff now.

**Comment by ismaeldcom** [comment]
After an internal conversation and some attempts, I keep seeing a strange behavior, but @tpaksu don't. We need a third one to try and confirm if it's on my side or a code related thing.

For me only happens from 8d4b0fbf51e4c6203eed178a72f33844bfc7558d, with fd7acbd85005fc6ed3354f179b3c487a706e5014 works OK.

Digging in the code, I found that commenting `$this->validate_settings();` in the last commit fix the problem, but don't know why this happens.

#### Steps to reproduce
- You need to have PayFast properly configured and be able to enable it
- Access [Payment methods settings](http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout)
- With the gateway disabled, click the toggle to enable
- Refresh
- Click the toggle to disable

Like you can see in the following video, when I click the toggle to disable after refresh it keeps enabled and send a request to enable instead of disable:

https://user-images.githubusercontent.com/7670276/121493093-c37f8d00-c9d7-11eb-8711-a695baae15d8.mov

cc @woocommerce/sigma

**Comment by dmallory42** [comment]
I tested with your instructions @ismaeldcom and I cannot reproduce that strange behaviour. For me, the following behaviour happens (which I think is expected):

- Enable PayFast
- Refresh the page, click the 'disable' toggle.
- PayFast is disabled with the `{"success":true,"data":false}` response returned from `admin-ajax.php`

**Comment by tpaksu** [comment]
As we discussed and agreed on making the service unavailable on the frontend instead of disabling it on backend when a configuration error is present @ p1623662651381200-slack-C01B8KNUYSW, the changes are now implemented and available for review.

**Comment by ismaeldcom** [comment]
When the setup is incomplete, it doesn't show up on checkout, as expected. But there are still some issues:

- Using blocks checkout, it's still available on incomplete setup.
- Still visible on checkout and blocks checkout when it's disabled.

For the later issue, I think you can check [another gateway code](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8a263ed73d475ee6bfe6442750715f585c638db9/includes/class-wc-gateway-stripe.php#L160-L166) they only return the false cases and then return `parent::is_available()` to not reimplement the enable/disable functionality.

**Review by ismaeldcom — APPROVED** [review]
LGTM now! Tested everything again and it's working as expected.

Thanks @tpaksu for working on it patiently and solve also the visibility of the payment method on both checkout screens when it is disabled.

Let's 🚢  it!
